CIP-0031/README.md

```diff
@@ -0,0 +1,210 @@
+
+## Motivation
+
+Datums in transaction outputs provide a way to store and access information on the blockchain.
+However, they are quite constrained in a number of ways.
+Most notably, in order to access the information which is contained in them, you have to _spend_ the output that the datum is attached to.
+
+This has a number of undesirable features:
+1. Even if the output is recreated after being spent, it is still a _new_ output. Any other user who wishes to look at the data cannot spend the old output (which is gone), but must rather spend the new output (which they will not know about until the next block). In practice this throttles some applications to one "operation" per block.
+2. Looking at the _information_ in the datum requires _spending_ the output, which means that care must be taken over the distribution of any funds in the output (possibly requiring scripts), and spending conditions must be met. This is overly stringent, inconvenient, and expensive.
+
+We would like to have a mechanism for accessing the information in datums that avoids these issues.
+
+### Use cases
+
+Here are some cases where we expect this to be helpful.
+
+1. Inspecting the state (datum, or locked value) of an on-chain application without having to consume the output, e.g. checking the current state of a stablecoin state machine.
+2. On-chain data providers that store data in outputs, to be referenced by other scripts.
+3. (With inline datums) Creating a single UTXO with data to be used in multiple subsequent transactions, but only paying the cost for submitting it once.
+
+### Enabling further improvements
+
+This proposal was designed in tandem with two further proposals which make use of reference scripts, CIP-32 (Inline datums) and CIP-33 (Reference scripts).
+A full assessment of the worth of this proposal should take into account the fact that it is an enabler for these other proposals.
+
+## Specification
+
+### Reference inputs
+
+We introduce a new kind of transaction input, a reference input.
+Transaction authors specify inputs as either normal (spending) inputs or reference inputs.[^1]
+
+A reference input is a transaction input, which is linked to a particular transaction output as normal, except that it _references_ the output rather than _spending_ it. That is:
+
+- A referenced output must exist in the UTXO set.
+- Any value on a referenced output is _not_ considered when balancing the transaction.
+- The spending conditions on referenced outputs are _not_ checked, nor are the witnesses required to be present. 
+    - i.e. validators are not required to pass (nor are the scripts themselves or redeemers required to be present at all), and signatures are not required for pubkey outputs.
+- Referenced outputs are _not_ removed from the UTXO set if the transaction validates.
+- Reference inputs _are_ visible to scripts.
```

Comment:
Together with CIP-33, reference inputs would then do two things: Be part of the context passed to scripts, and if they contain a reference script, that would get added to the witness set. This combination is a bit arbitrary. I worry about a future where we suddenly need to have different types of reference inputs because we need to be able to reference some of the things but not others.

Reply:
michaelpj:mpj/reference-inputs

Reply:
> Together with CIP-33, reference inputs would then do two things: Be part of the context passed to scripts, and if they contain a reference script, that would get added to the witness set. This combination is a bit arbitrary. I worry about a future where we suddenly need to have different types of reference inputs because we need to be able to reference some of the things but not others.

If a transaction's witnesses map contains additional (script hash, script source) pairs, does it really make a difference? All of the transaction's inputs were intentionally locked by specific script hashes, so they would never unintentionally refer to these additional scripts from the reference inputs...

Reply:
> Be part of the context passed to scripts, and if they contain a reference script, that would get added to the witness set.

My intention was that with CIP-33 _any_ input that corresponds to an output with a reference script would see it added to the witness set. So I claim that reference inputs still only do one (conceptual) thing: they let you _look at_ all the information in an output. It seems reasonable to me that looking at the information in an output that contains a reference script should let you use the reference script as a witness.

Rephrasing your worry, though, what you're suggesting is that we might want to e.g. restrict the information that a reference input lets us look at. I can't see a reason for that, but maybe there is one.